src/resolvers/relatedWork.ts

```diff
@@ -49,63 +82,55 @@ export const resolvers: Resolvers = {
-
-            if (newRelatedWork && !newRelatedWork.hasErrors()) {
-              // Version all of the plans (if any) and sync with the DMPHub
-              const plans = await Plan.findByProjectId(reference, context, project.id);
-              for (const plan of plans) {
-                await addVersion(context, plan, reference);
+          // Check if user has permission to modify project
+          const plan = await Plan.findById(reference, context, input.planId);
+          if (plan) {
+            const project = await Project.findById(reference, context, plan.projectId);
+            if (project && (await hasPermissionOnProject(context, project))) {
+              // Fetch or create work
+              let work = await Work.findByDoi(reference, context, input.doi);
+
+              if (!work) {
+                work = new Work({ doi: input.doi });
+                work = await work.create(context);
               }
-            }
-            return newRelatedWork;
-          }
-        }
-        throw context?.token ? ForbiddenError() : AuthenticationError();
-      } catch (err) {
-        if (err instanceof GraphQLError) throw err;
-
-        context.logger.error(prepareObjectForLogs(err), `Failure in ${reference}`);
-        throw InternalServerError();
-      }
-    },
 
-    // Update a related work on the research project
-    async updateRelatedWork(_, { input }, context: MyContext): Promise<RelatedWork> {
-      const reference = 'updateRelatedWork resolver';
-      try {
-        if (isAuthorized(context.token)) {
-          const relatedWork = await RelatedWork.findById(reference, context, input.relatedWorkId);
-          if (!relatedWork) {
-            throw NotFoundError();
-          }
-
-          const project = await Project.findById(reference, context, relatedWork.projectId);
-          if (project && await hasPermissionOnProject(context, project)) {
-            const toUpdate = new RelatedWork({ ...relatedWork, ...input });
-            const updated = await toUpdate.update(context);
+              // Fetch or create work version
+              let workVersion = await WorkVersion.findByDoiAndHash(reference, context, input.doi, input.hash);
+              if (!workVersion) {
+                // eslint-disable-next-line @typescript-eslint/no-unused-vars
+                const { planId, doi, ...options } = input;
+                workVersion = new WorkVersion(options);
+                workVersion.workId = work.id;
+                workVersion = await workVersion.create(context, work.doi);
+              }
 
-            if(updated && !updated.hasErrors()) {
-              // Version all of the plans (if any) and sync with the DMPHub
-              const plans = await Plan.findByProjectId(reference, context, project.id);
-              for (const plan of plans) {
-                await addVersion(context, plan, reference);
+              // Create related work
```

Comment:
You probably want to do another null check on `workVersion` here since it can be created above if it doesn't yet exist. Something like `if (!isNullOrUndefined(workVersion) && !workVersion.hasErrors())`

Reply:
Thanks, that's a good point.